### PR TITLE
Bind multiple collection to the same index

### PR DESCRIPTION
**Pull request by bidoubiwa**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -181,6 +181,37 @@ module.exports = {
...
+module.exports = {
+  indexName: "product"
+}
+```
+
+Nonetheless, it is not possible to know how many entries from each collection is added to MeiliSearch.
+
+For example, given two collections:
+- `Shoes`: with 300 entries and an `indexName` set to `clothes`
+- `Shirts`: 200 entries and an `indexName` set to `clothes`
```
Comment (bidoubiwa) [inline README.md:210]:
```suggestion
- `Shoes`: with 300 entries and an `indexName` set to `product`
- `Shirts`: 200 entries and an `indexName` set to `product`
```
```suggestion
- `Shoes`: with 300 entries and an `indexName` set to `clothes`
- `Shirts`: 200 entries and an `indexName` set to `clothes`
```

File: connectors/meilisearch/index.js [hunk 2]
```diff
@@ -114,33 +153,60 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
+
+          const indexed =
+            indexesNames.includes(indexUid) &&
+            indexedCollections.includes(collection)
+
+          // safe guard in case index does not exist anymore in MeiliSearch
+          if (
+            !indexesNames.includes(indexUid) &&
+            indexedCollections.includes(collection)
+          ) {
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:180]: put includes in variables

File: connectors/meilisearch/index.js [hunk 3]
```diff
@@ -114,33 +153,60 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
+          }
+
+          const {
+            numberOfDocuments = 0,
+            isIndexing = false,
+          } = indexesNames.includes(indexUid)
+            ? await this.getStats(indexUid)
+            : {}
+
+          const collectionsWithSameIndex = await collectionConnector.listCollectionsWithIndexName(
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:191]:
```suggestion
          const collectionsWithSameIndexName = await collectionConnector.listCollectionsWithIndexName(
```

File: connectors/meilisearch/index.js [hunk 4]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
      * @returns {number[]} - All updates id from the batched indexation process.
      */
     addCollectionInMeiliSearch: async function (collection) {
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:246]: To meilisearch

File: connectors/meilisearch/index.js [hunk 5]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
-        if (updateId) updateIds.push(updateId)
+        return updateId
       }
+
+      const updateIds = await this.actionInBatches(collection, addDocuments)
+      storeConnector.addIndexedCollection(collection)
       return { updateIds }
     },
 
+    emptyOrDeleteCollection: async function (collection) {
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:273]: emptyOrDeleteIndex

File: connectors/meilisearch/index.js [hunk 6]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
-      // Delete whole index only if the index is not a composite index
-      const indexUid = await collectionConnector.getIndexName(collection)
-
-      if (collection === collectionConnector.getIndexName(collection)) {
-        const { updateId } = await client.index(indexUid).deleteAllDocuments()
-        await this.waitForPendingUpdates({
-          updateId,
-          collection,
-        })
+      const indexedCollections = await storeConnector.getIndexedCollections()
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:300]: updateIndexInMeilISearch

File: connectors/meilisearch/index.js [hunk 7]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
     /**
-     * Remove a collection from MeiliSearch
+     * Apply an action on all the entries of the provided collection.
+     *
+     * @param  {string} collection
+     * @param  {function} callback - Function applied on each entry of the collection
+     *
+     * @returns {any[]} - List of all the returned elements from the callback.
+     */
+    actionInBatches: async function (collection, callback) {
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:315]: Switch to collection connector

File: connectors/meilisearch/index.js [hunk 8]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
+      return response
+    },
+    /**
+     * Search for the list of all collections that share the same index name.
+     *
+     * @param  {string} collection
+     *
+     * @returns {sring[]} - Collections names.
+     */
+    sameIndexCollections: async function (collection) {
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:341]: Naming bof

File: connectors/meilisearch/index.js [hunk 9]
```diff
@@ -168,72 +244,128 @@ module.exports = async ({ storeConnector, collectionConnector }) => {
...
-      const client = MeiliSearch({ apiKey, host })
-      const isCompositeIndex = collectionConnector.isCompositeIndex(collection)
-      const indexUid = await collectionConnector.getIndexName(collection)
-      if (!isCompositeIndex) {
-        await client.index(indexUid).delete()
-      } else {
-        // TODO if composite
-        await client.index(indexUid).delete()
-      }
+      await this.emptyOrDeleteCollection(collection)
```
Comment (bidoubiwa) [inline connectors/meilisearch/index.js:368]: Is it used?

File: package.json [hunk 10]
```diff
@@ -16,7 +16,7 @@
     "cy:run": "cypress run --env env=watch",
     "test": "jest",
     "test:watch": "jest --watch",
-    "test:e2e": "concurrently --kill-others -s first \"NODE_ENV=test yarn playground:soft_dev\" \"cypress run --env env=develop\"",
+    "test:e2e": "concurrently --kill-others -s first \"NODE_ENV=test yarn playground:soft_dev\" \"cypress open --env env=develop\"",
```
Comment (bidoubiwa) [inline package.json:19]:
```suggestion
    "test:e2e": "concurrently --kill-others -s first \"NODE_ENV=test yarn playground:soft_dev\" \"cypress run --env env=develop\"",
```
```suggestion
    "test:e2e": "concurrently --kill-others -s first \"NODE_ENV=test yarn playground:soft_dev\" \"cypress open --env env=develop\"",
```

**Review by bidoubiwa — COMMENTED** [review]
Review of @mdubus

**Review by mdubus — APPROVED** [review]
LGTM 🔥 
_(only small comments with no incidence on the code)_

**Comment by bidoubiwa** [comment]
bors merge

**Pull request merged**